spring-vault-core/src/main/java/org/springframework/vault/authentication/KubeServiceAccountTokenFile.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.vault.authentication;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.core.io.Resource;
+import org.springframework.util.StreamUtils;
+import org.springframework.vault.VaultException;
+
+/**
+ * Mechanism to retrieve a Kubernetes service account token.
+ * <p>
+ * A file containing a token for a pod’s service account is automatically mounted at
+ * <b>/var/run/secrets/kubernetes.io/serviceaccount/token</b>
+ *
+ * @author Michal Budzyn
+ * @see KubeJwtSupplier
+ */
+public class KubeServiceAccountTokenFile implements KubeJwtSupplier {
+
+	public static final String DEFAULT_KUBERNETES_SERVICE_ACCOUNT_TOKEN_FILE = "/var/run/secrets/kubernetes.io/serviceaccount/token";
+
+	private final Resource resource;
+
+	public KubeServiceAccountTokenFile() {
+		this(DEFAULT_KUBERNETES_SERVICE_ACCOUNT_TOKEN_FILE);
+	}
+
+	public KubeServiceAccountTokenFile(String fileName) {
+		this(new FileSystemResource(fileName));
+	}
+
+	public KubeServiceAccountTokenFile(File file) {
+		this(new FileSystemResource(file));
+	}
+
+	public KubeServiceAccountTokenFile(Resource resource) {
+		this.resource = resource;
+	}
+
+	@Override
+	public String getKubeJwt() {
+		try (InputStream is = resource.getInputStream()) {
```

Comment:
Is it safe to assume the service account token never changes when a pod is running?

Reply:
As far as I know token is not changed (see also [Service Account Key Rotation #20165](https://github.com/kubernetes/kubernetes/issues/20165)).

When you create a service account, a secret is created. This secret is mounted to the pod.   
In general secrets cannot be modified (only delete/recreate works).  

If you delete and recreate the serviceaccount a new secret will be created, but the old mount persist e.g. 

```
1. kubectl create serviceaccount hello
2. kubectl get serviceaccounts/hello -o yaml
kind: ServiceAccount
metadata:
  uid: 7b60565f-bf1f-11e7-b013-080027efc8f4
secrets:
- name: hello-token-rv80h

3. run hello-minikube with serviceaccount hello
4. kubectl get pods hello-minikube-1872458443-v4ld0 -o yaml

spec:
  containers:
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: hello-token-rv80h
      readOnly: true
  volumes:
  - name: hello-token-rv80h
    secret:
      defaultMode: 420
      secretName: hello-token-rv80h

5. kubectl delete serviceaccount hello && kubectl create serviceaccount hello
6. kubectl get serviceaccounts/hello -o yaml

kind: ServiceAccount
metadata:
  uid: 966fb602-bf25-11e7-b013-080027efc8f4
secrets:
- name: hello-token-h713z

7.  kubectl get pods hello-minikube-1872458443-v4ld0 -o yaml
kind: Pod
spec:
  containers:
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: hello-token-rv80h
      readOnly: true
  volumes:
  - name: hello-token-rv80h
    secret:
      defaultMode: 420
      secretName: hello-token-rv80h
```

Reply:
Thanks. That's good news in the light of `AuthenticationSteps`.

Reply:
You are welcome. If you don't mind, I would copy to this PR the changes I made in [Spring Cloud V #176](https://github.com/spring-cloud/spring-cloud-vault/pull/176) in the local_run_k8s.sh and minikube_ci_initialize.sh ?

Reply:
Feel free to update the PR in Spring Cloud Vault. Continuing the discussion in https://github.com/spring-cloud/spring-cloud-vault/pull/176.